src/Microsoft.OData.Core/Json/ODataUtf8JsonWriter.cs

```diff
@@ -797,34 +1063,138 @@ public async Task WriteValueAsync(sbyte value)
+            this.CommitUtf8JsonWriterContentsToBuffer();
+
+            WriteItemWithSeparatorIfNeeded();
+
+            this.bufferWriter.Write(doubleQuote.Slice(0, 1).Span);
+
+            await this.FlushAsync();
+
+            const int chunkSize = 2048;
+
+            int charsNotProcessedFromPreviousChunk = 0;
+
+            for (int i = 0; i < value.Length; i += chunkSize)
+            {
+                int remainingChars = Math.Min(chunkSize, value.Length - i);
+                bool isFinalBlock = (i + remainingChars) == value.Length;
+                ReadOnlyMemory<char> chunk = value.Slice(i - charsNotProcessedFromPreviousChunk, charsNotProcessedFromPreviousChunk + remainingChars);
+
+                int firstIndexToEscape = NeedsEscaping(chunk.Span);
+
+                Debug.Assert(firstIndexToEscape >= -1 && firstIndexToEscape < value.Length);
+
+                if (firstIndexToEscape != -1)
+                {
+                    WriteEscapedStringChunk(chunk.Span, firstIndexToEscape, isFinalBlock, out charsNotProcessedFromPreviousChunk);
+                }
+                else
+                {
+                    WriteStringChunk(chunk.Span, isFinalBlock);
+                }
+
+                // Flush the buffer if needed
+                await this.FlushIfBufferThresholdReachedAsync();
+            }
+
+            this.bufferWriter.Write(doubleQuote.Slice(0, 1).Span);
+
+            // since we bypass the Utf8JsonWriter, we need to signal to other
+            // Write methods that a separator should be written first
+            CheckIfSeparatorNeeded();
+
+            CheckIfAtManualValueArrayStart();
+        }
+
         public async Task WriteValueAsync(byte[] value)
         {
-            this.WriteSeparatorIfNecessary();
             if (value == null)
             {
+                this.WriteSeparatorIfNecessary();
                 this.utf8JsonWriter.WriteNullValue();
             }
+            else if (value.Length > bufferWriter.Capacity)
+            {
+                await WriteByteValueInChunksAsync(value);
+            }
             else
             {
+                this.WriteSeparatorIfNecessary();
                 this.utf8JsonWriter.WriteBase64StringValue(value);
             }
 
             await this.FlushIfBufferThresholdReachedAsync().ConfigureAwait(false);
         }
 
+        /// <summary>
+        /// Writes the byte value represented by the provided ReadOnlySpan in chunks using Base64 encoding.
+        /// </summary>
+        /// <param name="value">The ReadOnlySpan containing the byte value to be written.</param>
+        private async ValueTask WriteByteValueInChunksAsync(ReadOnlyMemory<byte> value)
+        {
+            this.CommitUtf8JsonWriterContentsToBuffer();
+
+            WriteItemWithSeparatorIfNeeded();
+
+            this.bufferWriter.Write(doubleQuote.Slice(0, 1).Span);
+
+            await FlushAsync().ConfigureAwait(false);
+
+            const int chunkSize = 4096;

```

Comment:
That said, even if it resizes, it would likely only happen once for the entire payload being written, and size the buffer is rented, the impact may be tolerable. So may not need to stress too much about it. I would suggest running the allocation benchmarks with different chunk sizes and pick one. We can change it later if we get more real world data. Some overhead is to be expected when writing large values.

Reply:
Used a chunksize of `2048`